### PR TITLE
Automatic savefile upgrade

### PR DESCRIPTION
**Pull request by Coding-Hen**
Includes the changes from #1517 and upgrades the file to the latest version upon startup to prevent wiping of original save and creation of a new world

### CODE REVIEW

File: NitroxServer/Serialization/World/WorldPersistence.cs [hunk 1]
```diff
@@ -214,5 +213,49 @@ public World CreateWorld(PersistedWorldData pWorldData, ServerGameMode gameMode)
 
             return world;
         }
+
+        private void UpgradeSave()
+        {
+            string saveDir = config.SaveName;
```
Comment (Jannify) [inline NitroxServer/Serialization/World/WorldPersistence.cs:219]: Would make `saveDir` a parameter just to be safe

File: NitroxServer/ConsoleCommands/UpgradeCommand.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
Comment (Jannify) [inline NitroxServer/ConsoleCommands/UpgradeCommand.cs:1]: I don't see the reason to keep this command when we do it on load automatically

File: NitroxServer/Serialization/World/WorldPersistence.cs [hunk 3]
```diff
@@ -214,5 +213,49 @@ public World CreateWorld(PersistedWorldData pWorldData, ServerGameMode gameMode)
...
+            // SaveFileVersion structure was updated in V1.5.0.0
+            // This can be removed with V1.6.0.0 or later
+            if (File.ReadAllText(Path.Combine(saveDir, $"Version{FileEnding}")).Contains("BaseDataVersion"))
+            {
+                saveFileVersion = new SaveFileVersion(new Version(1, 4, 0, 0));
+            }
+
+            if (saveFileVersion.Version == NitroxEnvironment.Version)
+            {
+                Log.Info("Save files are already at the newest version");
```
Comment (Jannify) [inline NitroxServer/Serialization/World/WorldPersistence.cs:232]: Don't know if we should keep this Log. Would be executed on every server start...

**Review by Jannify — CHANGES_REQUESTED** [review]

File: NitroxServer-Subnautica/Program.cs [hunk 4]
```diff
@@ -37,6 +37,7 @@ private static async Task Main(string[] args)
             AppDomain.CurrentDomain.UnhandledException += CurrentDomainOnUnhandledException;
 
             ConfigureCultureInfo();
+            Log.Info($"Starting NitroxServer V{NitroxEnvironment.Version} for Subnautica");
```
Comment (dartasen) [inline NitroxServer-Subnautica/Program.cs:40]:
```suggestion
            Log.Info($"Starting NitroxServer V{NitroxEnvironment.Version}");
```

File: NitroxServer/ConsoleCommands/UpgradeCommand.cs [hunk 5]
```diff
@@ -34,6 +34,13 @@ protected override void Execute(CallArgs args)
 
             SaveFileVersion saveFileVersion = worldPersistence.Serializer.Deserialize<SaveFileVersion>(Path.Combine(saveDir, $"Version{fileEnding}"));
 
+            // SaveFileVersion structure was updated in V1.5.0.0
+            // This can be removed with V1.6.0.0 or later
```
Comment (dartasen) [inline NitroxServer/ConsoleCommands/UpgradeCommand.cs:38]: Is this something worth keeping ? I do think it's not worth it maintaining save files at this stage of the developement, imo upgrades would fit better for future version, once the core mod is stabilized

File: NitroxServer/Serialization/World/WorldPersistence.cs [hunk 6]
```diff
@@ -214,5 +213,49 @@ public World CreateWorld(PersistedWorldData pWorldData, ServerGameMode gameMode)
...
         }
+
+        private void UpgradeSave()
+        {
+            string saveDir = config.SaveName;
+
+            SaveFileVersion saveFileVersion = Serializer.Deserialize<SaveFileVersion>(Path.Combine(saveDir, $"Version{FileEnding}"));
+
+            // SaveFileVersion structure was updated in V1.5.0.0
+            // This can be removed with V1.6.0.0 or later
```
Comment (dartasen) [inline NitroxServer/Serialization/World/WorldPersistence.cs:224]: same

File: NitroxServer/Serialization/World/WorldPersistence.cs [hunk 7]
```diff
@@ -91,12 +94,8 @@ internal Optional<World> LoadFromFile(string saveDir)
             try
             {
                 PersistedWorldData persistedData = new();
-                SaveFileVersion saveFileVersion = Serializer.Deserialize<SaveFileVersion>(Path.Combine(saveDir, $"Version{FileEnding}"));
 
-                if (saveFileVersion == null || saveFileVersion.Version != NitroxEnvironment.Version)
-                {
-                    throw new InvalidDataException("Version file is empty or save data files are too old");
-                }
```
Comment (dartasen) [inline NitroxServer/Serialization/World/WorldPersistence.cs:99]: I think those kind of data are good to keep, maybe just a log

File: NitroxServer-Subnautica/Program.cs (hunk 4 above)
Comment (Jannify) [reply NitroxServer-Subnautica/Program.cs:40]: As this is in `NitroxServer-Subnautica` I saw this fitting. Especially with BZ in mind.

File: NitroxServer/Serialization/World/WorldPersistence.cs (hunk 7 above)
Comment (Jannify) [reply NitroxServer/Serialization/World/WorldPersistence.cs:99]: Yeah the `saveFileVersion == null` check should be kept imo.

File: NitroxServer/ConsoleCommands/UpgradeCommand.cs (hunk 5 above)
Comment (Jannify) [reply NitroxServer/ConsoleCommands/UpgradeCommand.cs:38]: I would leave it in for this version. It's just a couple lines of extra code and the upgrade script is already written. On top of that we could use it as a small testing/beta phase on the users as only a few of them were using JSON as save-file-format.

File: NitroxServer/Serialization/World/WorldPersistence.cs (hunk 3 above)
Comment (Coding-Hen) [reply NitroxServer/Serialization/World/WorldPersistence.cs:232]: Yes I agree, it was something I was thinking about removing as you are right does show each startup.

File: NitroxServer/Serialization/World/WorldPersistence.cs (hunk 7 above)
Comment (Coding-Hen) [reply NitroxServer/Serialization/World/WorldPersistence.cs:99]: Wont this get caught by the check of is the file version.json exists? I believe the removed saveFileVersion with deserialization defaults to version 1.5.0.0 when it cannot deserialise the content so should never be null

Comment (Jannify) [reply NitroxServer/Serialization/World/WorldPersistence.cs:99]: Even if the file does not exist or is corrupted?

Comment (Coding-Hen) [reply NitroxServer/Serialization/World/WorldPersistence.cs:99]: If it doesnt exist then the [check](https://github.com/SubnauticaNitrox/Nitrox/blob/master/NitroxServer/Serialization/World/WorldPersistence.cs#L85) for a previous file should catch it unless I am mistaken. If it is currupted like for example a 1.4 file with the BaseDataVersion then it defaults to 1.5.0.0

**Review by Jannify — APPROVED** [review]

**Pull request merged**